packages/devtools_app/lib/src/timeline/flutter/timeline_flame_chart.dart

```diff
@@ -1,17 +1,349 @@
+      });
+    }
+    super.didUpdateWidget(oldWidget);
+  }
+
+  @override
+  void initState() {
+    super.initState();
+
+    // TODO(kenz): improve this so we are not rebuilding on every scroll.
+    _scrollControllerX = ScrollController()
+      ..addListener(() {
+        setState(() {
+          scrollOffsetX = _scrollControllerX.offset;
+        });
+      });
+
+    _scrollControllerY = ScrollController()
+      ..addListener(() {
+        setState(() {
+          scrollOffsetY = _scrollControllerY.offset;
+        });
+      });
+  }
+
+  @override
+  void dispose() {
+    _scrollControllerX.dispose();
+    _scrollControllerY.dispose();
+    _selectedEventSubscription.cancel();
+    // TODO(kenz): dispose [_controller] here.
+    super.dispose();
+  }
+
   @override
   Widget build(BuildContext context) {
-    return Container(
-      color: Colors.black26,
-      child: const Center(
-        child: Text('TODO Flame Chart'),
-      ),
+    return LayoutBuilder(
+      builder: (context, constraints) {
+        return Scrollbar(
+          child: SingleChildScrollView(
+            controller: _scrollControllerX,
+            scrollDirection: Axis.horizontal,
+            child: Scrollbar(
+              child: SingleChildScrollView(
+                controller: _scrollControllerY,
+                scrollDirection: Axis.vertical,
+                child: _flameChartBody(constraints),
+              ),
+            ),
+          ),
+        );
+      },
+    );
+  }
+
+  Widget _flameChartBody(BoxConstraints constraints) {
+    final width = math.max(constraints.maxWidth, widget.totalStartingWidth);
+    final height = math.max(constraints.maxHeight, widget.height);
+
+    return Stack(
+      children: [
+        Container(
+          width: width,
+          height: height,
+        ),
+        ..._nodesInViewport(constraints), // pick what to show
+      ],
+    );
+  }
+
+  List<FlameChartNode> _nodesInViewport(BoxConstraints constraints) {
+    // TODO(kenz): is creating all the FlameChartNode objects expensive even if
+    // we won't add them to the view? We create all the FlameChartNode objects
+    // and place them in FlameChart rows, but we only add [nodesInViewport] to
+    // the widget tree.
+    _buildFlameChartElements();
+
+    final nodesInViewport = <FlameChartNode>[];
+    for (var row in rows) {
+      for (var node in row.nodes) {
+        final fitsHorizontally = node.rect.right >= scrollOffsetX &&
+            node.rect.left - scrollOffsetX <= constraints.maxWidth;
+        final fitsVertically = node.rect.bottom >= scrollOffsetY &&
+            node.rect.top - scrollOffsetY <= constraints.maxHeight;
+        if (fitsHorizontally && fitsVertically) {
+          nodesInViewport.add(node);
+        }
+      }
+    }
+    return nodesInViewport;
+  }
+
+  void _buildFlameChartElements() {
+    _resetColorOffsets();
+
+    rows
+      ..clear()
+      ..addAll(List.generate(
```

Comment:
this code is on the critical path so just use straight imperative code instead of addAll and List.generate

Reply:
done